mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala

```diff
@@ -218,31 +217,51 @@ class LogisticRegression(override val uid: String)
 
   override def getThreshold: Double = super.getThreshold
 
+  /**
+   * Whether to over-/undersamples each of training sample according to the given
+   * weight in `weightCol`. If empty, all samples are supposed to have weights as 1.0.
+   * Default is empty, so all samples have weight one.
+   * @group setParam
+   */
+  def setWeightCol(value: String): this.type = set(weightCol, value)
+  setDefault(weightCol -> "")
+
   override def setThresholds(value: Array[Double]): this.type = super.setThresholds(value)
 
   override def getThresholds: Array[Double] = super.getThresholds
 
   override protected def train(dataset: DataFrame): LogisticRegressionModel = {
     // Extract columns from data.  If dataset is persisted, do not persist oldDataset.
-    val instances = extractLabeledPoints(dataset).map {
-      case LabeledPoint(label: Double, features: Vector) => (label, features)
+    val instances: RDD[(Double, Double, Vector)] = if ($(weightCol).isEmpty) {
```

Comment:
I didn't check the previous version. I would do the following:

~~~scala
val w = if ($(weightCol).isEmpty) lit(1.0) else col($(weightCol))
dataset.select(col($(label)), w, col($featuresCol)).map {
  case Row(label: Double, weight: Double, features: Vector) =>
    (label, weight, features)
  }
}
~~~

Btw, if it helps code readability, we can defined a private case class called `WeightedLabelPoint` or `Instance` in `LogisticRegression`.